util/rpcclient/rpcclient.go

```diff
@@ -0,0 +1,208 @@
+package rpcclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	flag "github.com/spf13/pflag"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/offchainlabs/nitro/util/signature"
+)
+
+type ClientConfig struct {
+	URL            string        `koanf:"url"`
+	JWTSecret      string        `koanf:"jwtsecret"`
+	Timeout        time.Duration `koanf:"timeout" reload:"hot"`
+	Retries        uint          `koanf:"retries" reload:"hot"`
+	ConnectionWait time.Duration `koanf:"connection-wait"`
+	ArgLogLimit    uint          `koanf:"arg-log-limit" reload:"hot"`
+	RetryErrors    string        `koanf:"retry-errors" reload:"hot"`
+}
+
+type ClientConfigFetcher func() *ClientConfig
+
+var TestClientConfig = ClientConfig{
+	URL:       "self",
+	JWTSecret: "",
+}
+
+var DefaultClientConfig = ClientConfig{
+	URL:         "self-auth",
+	JWTSecret:   "",
+	ArgLogLimit: 2048,
+}
+
+func RPCClientAddOptions(prefix string, f *flag.FlagSet, defaultConfig *ClientConfig) {
+	f.String(prefix+".url", defaultConfig.URL, "url of server, use self for loopback websocket, self-auth for loopback with authentication")
+	f.String(prefix+".jwtsecret", defaultConfig.JWTSecret, "path to file with jwtsecret for validation - ignored if url is self or self-auth")
+	f.Duration(prefix+".connection-wait", defaultConfig.ConnectionWait, "how long to wait for initial connection")
+	f.Duration(prefix+".timeout", defaultConfig.Timeout, "per-response timeout (0-disabled)")
+	f.Uint(prefix+".arg-log-limit", defaultConfig.ArgLogLimit, "limit size of arguments in log entries")
+	f.Uint(prefix+".retries", defaultConfig.Retries, "number of retries in case of failure(0 mean one attempt)")
+	f.String(prefix+".retry-errors", defaultConfig.RetryErrors, "Errors matching this regular expression are automatically retried")
+}
+
+type RpcClient struct {
+	config    ClientConfigFetcher
+	client    *rpc.Client
+	autoStack *node.Node
+	logId     uint64
+}
+
+func NewRpcClient(config ClientConfigFetcher, stack *node.Node) *RpcClient {
+	return &RpcClient{
+		config:    config,
+		autoStack: stack,
+	}
+}
+
+func (c *RpcClient) Close() {
+	c.client.Close()
+}
+
+func limitString(limit int, str string) string {
+	if limit == 0 || len(str) <= limit {
+		return str
+	}
+	prefix := str[:limit/2-1]
+	postfix := str[len(str)-limit/2+1:]
+	return fmt.Sprintf("%v...%v", prefix, postfix)
+}
+
+func logArgs(limit int, args ...interface{}) string {
+	res := "["
+	for i, arg := range args {
+		res += limitString(limit, fmt.Sprintf("%+v", arg))
```

Comment:
We should consider logging the JSON serialized form of the argument instead of the argument itself. This can matter with things like CallMsg, and logging the JSON serialized form makes it easier to test out manually against an L1 node.